apollo-router/src/query_planner/fetch.rs

```diff
@@ -243,6 +255,189 @@ impl Display for QueryHash {
+
+            let as_mut = new.make_mut();
+
+            as_mut.arguments.iter_mut().for_each(|arg| {
+                if arg.value.as_variable() == Some(&from) {
+                    arg.make_mut().value = ast::Value::Variable(to.clone()).into();
+                }
+            });
+
+            as_mut.selection_set = as_mut
+                .selection_set
+                .clone()
+                .into_iter()
+                .map(|s| rename_variables(s, from.clone(), to.clone()))
+                .collect();
+
+            ast::Selection::Field(new)
+        }
+        ast::Selection::InlineFragment(f) => {
+            let mut new = f.clone();
+            new.make_mut().selection_set = f
+                .selection_set
+                .clone()
+                .into_iter()
+                .map(|s| rename_variables(s, from.clone(), to.clone()))
+                .collect();
+            ast::Selection::InlineFragment(new)
+        }
+        ast::Selection::FragmentSpread(f) => ast::Selection::FragmentSpread(f),
+    }
+}
+
+#[test]
+fn test_query_batching_for_contextual_args() {
+    let old_query = "query QueryLL__Subgraph1__1($representations:[_Any!]!$Subgraph1_U_field_a:String!){_entities(representations:$representations){...on U{id field(a:$Subgraph1_U_field_a)}}}";
+    let mut ctx_args = HashSet::new();
+    ctx_args.insert("Subgraph1_U_field_a".to_string());
+    let contextual_args = Some((ctx_args, 2));
+
+    let expected = "query QueryLL__Subgraph1__1($representations: [_Any!]!, $Subgraph1_U_field_a_0: String!, $Subgraph1_U_field_a_1: String!) { _0: _entities(representations: $representations) { ... on U { id field(a: $Subgraph1_U_field_a_0) } } _1: _entities(representations: $representations) { ... on U { id field(a: $Subgraph1_U_field_a_1) } } }";
+
+    assert_eq!(
+        expected,
+        query_batching_for_contextual_args(old_query, &contextual_args)
+            .unwrap()
+            .unwrap()
+    );
+}
+
+// TODO: There is probably a function somewhere else that already does this
+fn data_at_path<'v>(data: &'v Value, path: &Path) -> Option<&'v Value> {
+    let v = match &path.0[0] {
+        PathElement::Fragment(s) => {
+            // get the value at data.get("__typename") and compare it with s. If the values are equal, return data, otherwise None
+            let mut z: Option<&Value> = None;
+            let wrapped_typename = data.get("__typename");
+            if let Some(t) = wrapped_typename {
+                if t.as_str() == Some(s.as_str()) {
+                    z = Some(data);
+                }
+            }
+            z
+        }
+        PathElement::Key(v, _) => data.get(v),
+        PathElement::Index(idx) => Some(&data[idx]),
+        PathElement::Flatten(_) => None,
+    };
+
+    if path.len() > 1 {
+        if let Some(val) = v {
+            let remaining_path = path.iter().skip(1).cloned().collect();
+            return data_at_path(val, &remaining_path);
+        }
+    }
+    v
+}
```

Comment:
I don't think this exists yet, but I think this can perhaps be a (`pub(crate) fn`) getter on the `Path` struct.

Reply:
I think because there's custom notation in here, I'll leave it where it is and remove the TODO.